README.md
CCNx Simple Ping
====

This is a trivial ping based on the Hello World.   It is just a simple example,
it is not meant to be complete or correct.

Copyright 2015, Palo Alto Research Center Inc.



Building
-----

This code assumes an Ubuntu 14.04 with the CCNx binary distribution installed.
It is meant to show a simple example of a makefile and the libraries required
to build a simple piece of software.


Running
----
* Run the forwarder (metis)
* Run the ccnx-ping-server
* Run ccnx-ping

Limitations
----
This code doesn't have the ability to select the name being pinged on the
command line.  It's just a test for a single forwarder.

Contact
----

You can find more information at the ccnx website: http://www.ccnx.org/


License
----

This software is licensed under a BSD license. See LICENSE file.


ChangeLog
# Change Log
## [1.0.20150518] - 2015-05-18
### Changed
- Initial, simple release



ccnx-ping.c
/*
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 * Copyright 2014-2015 Palo Alto Research Center, Inc. (PARC), a Xerox company.  All Rights Reserved.
 * The content of this file, whole or in part, is subject to licensing terms.
 * If distributing this software, include this License Header Notice in each
 * file and provide the accompanying LICENSE file.
 */
/**
 * @author Glenn Scott, Computing Science Laboratory, PARC
 * @copyright 2014-2015 Palo Alto Research Center, Inc. (PARC), A Xerox Company. All Rights Reserved.
 */
#include <config.h>
#include <stdio.h>

#include <LongBow/runtime.h>

#include <ccnx/api/ccnx_Portal/ccnx_Portal.h>
#include <ccnx/api/ccnx_Portal/ccnx_PortalRTA.h>

#include <parc/security/parc_Security.h>
#include <parc/security/parc_IdentityFile.h>
#include <parc/security/parc_PublicKeySignerPkcs12Store.h>

#include <parc/algol/parc_Memory.h>

PARCIdentity *
createAndGetIdentity(void)
{
    const char *keystoreName = "consumer_keystore";
    const char *keystorePassword = "keystore_password";
    unsigned int keyLength = 1024;
    unsigned int validityDays = 30;
    char *subjectName = "consumer";

    bool success = parcPublicKeySignerPkcs12Store_CreateFile(keystoreName, keystorePassword, subjectName, keyLength, validityDays);
    assertTrue(success,
               "parcPublicKeySignerPkcs12Store_CreateFile('%s', '%s', '%s', %d, %d) failed.",
               keystoreName, keystorePassword, subjectName, keyLength, validityDays);

    PARCIdentityFile *identityFile = parcIdentityFile_Create(keystoreName, keystorePassword);
    PARCIdentity *identity = parcIdentity_Create(identityFile, PARCIdentityFileAsPARCIdentity);
    parcIdentityFile_Release(&identityFile);

    return identity;
}

CCNxPortalFactory *
setupConsumerFactory(void)
{
    PARCIdentity *identity = createAndGetIdentity();

    CCNxPortalFactory *factory = ccnxPortalFactory_Create(identity);
    parcIdentity_Release(&identity);
    return factory;
}

int
consumer(void)
{
    parcSecurity_Init();
    
    CCNxPortalFactory *factory = setupConsumerFactory();
   
    CCNxPortal *portal = ccnxPortalFactory_CreatePortal(factory, ccnxPortalRTA_Message, &ccnxPortalAttributes_Blocking);

    assertNotNull(portal, "Expected a non-null CCNxPortal pointer.");

    CCNxName *name = ccnxName_CreateFromURI("lci:/Ping/World");

    CCNxInterest *interest = ccnxInterest_CreateSimple(name);
    ccnxName_Release(&name);

    CCNxMetaMessage *message = ccnxMetaMessage_CreateFromInterest(interest);

    struct timeval send_time;
    struct timeval receive_time;
	int ret;

	long latency       = 0;
	long total_latency = 0;
	long iterations    = 0;


	while(true){
	  gettimeofday(&send_time, NULL);
	  if (ccnxPortal_Send(portal, message)) {
		  while (ccnxPortal_IsError(portal) == false) {
			  CCNxMetaMessage *response = ccnxPortal_Receive(portal);
			  gettimeofday(&receive_time, NULL);
			  if (response != NULL) {
				  if (ccnxMetaMessage_IsContentObject(response)) {
					  CCNxContentObject *contentObject = ccnxMetaMessage_GetContentObject(response);

					  latency = 
						  (receive_time.tv_sec - send_time.tv_sec)*1000000 +
						  (receive_time.tv_usec - send_time.tv_usec);
					  total_latency = total_latency + latency;
			          iterations = iterations + 1;
					  printf("%6lu us  (%6lu us avg)\n", latency, total_latency/iterations);
						  
					  //PARCBuffer *payload = ccnxContentObject_GetPayload(contentObject);

					  //char *string = parcBuffer_ToString(payload);
					  //printf("%s\n", string);
					  //parcMemory_Deallocate((void **)&string);

					  break;
				  }
			  }
			  ccnxMetaMessage_Release(&response);
		  }
	  }
	  sleep(1);
    }

    ccnxPortal_Release(&portal);

    ccnxPortalFactory_Release(&factory);
    
    parcSecurity_Fini();
    return 0;
}

int
main(int argc, char *argv[argc])
{
    return consumer();
}


helloWorld_ConsumerForever.c
/*
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 * Copyright 2014 Palo Alto Research Center, Inc. (PARC), a Xerox company.  All Rights Reserved.
 * The content of this file, whole or in part, is subject to licensing terms.
 * If distributing this software, include this License Header Notice in each
 * file and provide the accompanying LICENSE file.
 */
/**
 * @author Glenn Scott, Computing Science Laboratory, PARC
 * @copyright 2014 Palo Alto Research Center, Inc. (PARC), A Xerox Company. All Rights Reserved.
 */
#include <config.h>
#include <stdio.h>

#include <LongBow/runtime.h>

#include <ccnx/api/ccnx_Portal/ccnx_Portal.h>
#include <ccnx/api/ccnx_Portal/ccnx_PortalRTA.h>

#include <parc/security/parc_Security.h>
#include <parc/security/parc_IdentityFile.h>
#include <parc/security/parc_PublicKeySignerPkcs12Store.h>

#include <parc/algol/parc_Memory.h>

PARCIdentity *
createAndGetIdentity(void)
{
    parcSecurity_Init();

    const char *keystoreName = "consumer_keystore";
    const char *keystorePassword = "keystore_password";
    unsigned int keyLength = 1024;
    unsigned int validityDays = 30;
    char *subjectName = "consumer";

    bool success = parcPublicKeySignerPkcs12Store_CreateFile(keystoreName, keystorePassword, subjectName, keyLength, validityDays);
    assertTrue(success,
               "parcPublicKeySignerPkcs12Store_CreateFile('%s', '%s', '%s', %d, %d) failed.",
               keystoreName, keystorePassword, subjectName, keyLength, validityDays);

    PARCIdentityFile *identityFile = parcIdentityFile_Create(keystoreName, keystorePassword);
    PARCIdentity *identity = parcIdentity_Create(identityFile, PARCIdentityFileAsPARCIdentity);
    parcIdentityFile_Release(&identityFile);

    parcSecurity_Fini();

    return identity;
}

CCNxPortalFactory *
setupConsumerFactory(void)
{
    PARCIdentity *identity = createAndGetIdentity();

    CCNxPortalFactory *factory = ccnxPortalFactory_Create(identity);
    parcIdentity_Release(&identity);
    return factory;
}

int
consumer(void)
{
    CCNxPortalFactory *factory = setupConsumerFactory();

    CCNxPortal *portal =
        ccnxPortalFactory_CreatePortal(factory, ccnxPortalRTA_Message, &ccnxPortalAttributes_Blocking);

    assertNotNull(portal, "Expected a non-null CCNxPortal pointer.");
    
    CCNxName *name = ccnxName_CreateFromURI("lci:/Hello/World");
    CCNxInterest *interest = ccnxInterest_CreateSimple(name);
    ccnxName_Release(&name);
    
    for (int i = 0; i < 50000; i++) {
        CCNxMetaMessage *message = ccnxMetaMessage_CreateFromInterest(interest);
        
        if (ccnxPortal_Send(portal, message)) {
            while (ccnxPortal_IsError(portal) == false) {
                CCNxMetaMessage *response = ccnxPortal_Receive(portal);
                if (response != NULL) {
                    if (ccnxMetaMessage_IsContentObject(response)) {
                        break;
                    }
                }
                ccnxMetaMessage_Release(&response);
            }
        }
    }

    ccnxPortal_Release(&portal);

    ccnxPortalFactory_Release(&factory);

    return 0;
}

int
main(int argc, char *argv[argc])
{
    return consumer();
}


ccnx-ping-server.c
/*
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 * Copyright 2014-2015 Palo Alto Research Center, Inc. (PARC), a Xerox company.  All Rights Reserved.
 * The content of this file, whole or in part, is subject to licensing terms.
 * If distributing this software, include this License Header Notice in each
 * file and provide the accompanying LICENSE file.
 */
/**
 * @author Glenn Scott, Computing Science Laboratory, PARC
 * @copyright 2014-2015 Palo Alto Research Center, Inc. (PARC), A Xerox Company. All Rights Reserved.
 */
#include <config.h>
#include <LongBow/runtime.h>

#include <time.h>

#include <ccnx/api/ccnx_Portal/ccnx_Portal.h>
#include <ccnx/api/ccnx_Portal/ccnx_PortalRTA.h>

#include <parc/security/parc_Security.h>
#include <parc/security/parc_PublicKeySignerPkcs12Store.h>
#include <parc/security/parc_IdentityFile.h>

#include <ccnx/common/ccnx_Name.h>

PARCIdentity *
createAndGetIdentity(void)
{
    const char *keystoreName = "producer_keystore";
    const char *keystorePassword = "keystore_password";
    unsigned int keyLength = 1024;
    unsigned int validityDays = 30;
    char *subjectName = "producer";

    bool success = parcPublicKeySignerPkcs12Store_CreateFile(keystoreName, keystorePassword, subjectName, keyLength, validityDays);
    assertTrue(success,
               "parcPublicKeySignerPkcs12Store_CreateFile('%s', '%s', '%s', %d, %d) failed.",
               keystoreName, keystorePassword, subjectName, keyLength, validityDays);

    PARCIdentityFile *identityFile = parcIdentityFile_Create(keystoreName, keystorePassword);
    PARCIdentity *identity = parcIdentity_Create(identityFile, PARCIdentityFileAsPARCIdentity);
    parcIdentityFile_Release(&identityFile);

    return identity;
}

CCNxPortalFactory *
setupPortalFactory(void)
{
    PARCIdentity *identity = createAndGetIdentity();

    CCNxPortalFactory *factory = ccnxPortalFactory_Create(identity);
    parcIdentity_Release(&identity);
    return factory;
}

PARCBuffer *
makePayload(void)
{
    time_t theTime = time(0);

    PARCBufferComposer *composer = parcBufferComposer_Create();
    parcBufferComposer_Format(composer, "Ping World. The time is %s", ctime(&theTime));
    PARCBuffer *payload = parcBufferComposer_ProduceBuffer(composer);
    parcBufferComposer_Release(&composer);

    return payload;
}

int
producer(void)
{
    parcSecurity_Init();

    CCNxPortalFactory *factory = setupPortalFactory();

    CCNxPortal *portal = ccnxPortalFactory_CreatePortal(factory, ccnxPortalRTA_Message, &ccnxPortalAttributes_Blocking);
    
    assertNotNull(portal, "Expected a non-null CCNxPortal pointer.");

    CCNxName *listenName = ccnxName_CreateFromURI("lci:/Ping");
    CCNxName *goodbye = ccnxName_CreateFromURI("lci:/Ping/Goodbye%21");
    CCNxName *contentName = ccnxName_CreateFromURI("lci:/Ping/World");

    if (ccnxPortal_Listen(portal, listenName)) {
        while (true) {
            CCNxMetaMessage *request = ccnxPortal_Receive(portal);

            if (request == NULL) {
                break;
            }

            CCNxInterest *interest = ccnxMetaMessage_GetInterest(request);

            if (interest != NULL) {
                CCNxName *interestName = ccnxInterest_GetName(interest);

                if (ccnxName_Equals(interestName, contentName)) {

                    PARCBuffer *payload = makePayload();

                    CCNxContentObject *contentObject = ccnxContentObject_CreateWithDataPayload(contentName, payload);

                    CCNxMetaMessage *message = ccnxMetaMessage_CreateFromContentObject(contentObject);

                    if (ccnxPortal_Send(portal, message) == false) {
                        fprintf(stderr, "ccnxPortal_Send failed: %d\n", ccnxPortal_GetError(portal));
                    }

                    ccnxMetaMessage_Release(&message);

                    parcBuffer_Release(&payload);
                } else if (ccnxName_Equals(interestName, goodbye)) {
                    break;
                }
            }
            ccnxMetaMessage_Release(&request);
        }
    }

    ccnxName_Release(&listenName);
    ccnxName_Release(&goodbye);
    ccnxName_Release(&contentName);

    ccnxPortal_Release(&portal);

    ccnxPortalFactory_Release(&factory);

    parcSecurity_Fini();

    return 0;
}

int
main(int argc, const char *argv[argc])
{
    return producer();
}
